### DISCUSSION TITLE
Simulationsmodus für Skripte

### DISCUSSION THREAD

**Discussion by chef76** [brutella/hkknx-public#113] [Ideas]
Hallo Matthias,

Ich kämpfe noch ein bisschen mit meiner temperaturabhängigen Lamellensteuerung.

Wenn es einen Debug-/Simulationsmodus gäbe, wäre das Problem längst gelöst.

Mein eibPort hat einen Simulationsmodus, wo du das Skript laufen lassen und dann die Eingänge (Variablen) von Hand setzen kannst. Man sieht dann optisch sofort, was die Ausgabevariablen machen. Es wird dabei weder vom Bus gelesen, noch gesendet. Du könntest die Variablenwerte entweder per Mousover oder dauerhaft neben der Variablen anzeigen. Als Krönung geht beim eibPort auch eine Zeitsimulation, wenn z.B. zeitabhängige Komponenten im Skript sind. Die Zeit kann man dann von Echtzeit bis auf den Faktor 10000 beschleunigen.

Findet das außer mir noch jemand interessant?

**Comment by brutella** [comment]
Ich weiß jetzt nicht im Detail wie dein Skript aussieht, aber du könntest mit relativ wenig Aufwand Werte simulieren.
Nehmen wir an deine Automation empfängt Temperaturwerte und steuert abhängig vom Temperaturwert.

```go
var ch = hkknx.GroupWriteNotify("0/0/1")
buf = <- ch
temp = hkknx.ParseDPT9001(buf)

if temp > 23.5 {
    // Lamellen verstellen
}
```

1. Zum Testen würde ich sowieso immer Eingangsvariable mit `println` oder `printf` in die Konsole ausgeben. Somit sieht man schnell, ob der empfangene Wert mit dem erwarteten Wert übereinstimmt.

```go
println("temp =", temp)
```

2. Wenn du einen Temperaturwert simulieren möchtest, dann kann du mit einem zweiten Skript eine beliebigen Wert an die Gruppenadresse senden. Erstelle dazu ein zweites Skript, füge den Code ein und klicke auf *Ausführen*.

```go
hkknx.GroupWrite(hkknx.DPT9001(23.4), "0/0/1")
```

Für Testzwecke wäre es auch ratsam eine unbenutzte Gruppenadresse verwenden. Wenn die Automation dann soweit getestet ist, kann man die richtige Gruppenadresse verwenden.

Was hältst du davon?

**Reply by chef76** [reply]
Danke für die Lösungsvorschläge.

So weit bin ich schon, ich hab mir schon Test-GAs angelegt und diese in HomeKit eingebunden. Die zeigen mir dann Live die Zustände und ich kann auch Werte senden. Das ist halt maximal aufwändig. 

Mit dem Simulationsmodus meines eibPort ist das mit Null Zusatzaufwand und ohne Gefrickel am Code möglich.

Minimalanforderung wäre aus meiner Sicht, dass man im Skript per Mouseover die aktuellen Zustände der Variablen sehen kann. Das sollte doch leicht umsetzbar sein?

Ich bin jetzt wahrlich kein Programmierer, aber ich denke die Brücke nutzen auch Leute, die da noch weniger bewandert sind als ich. Mit Homebridge und Co. ist das ungefähr derselbe Aufwand und willst Du da nicht besser sein? :-)

**Reply by brutella** [reply]
> Minimalanforderung wäre aus meiner Sicht, dass man im Skript per Mouseover die aktuellen Zustände der Variablen sehen kann. Das sollte doch leicht umsetzbar sein?

Ist leider nicht so einfach wie man denkt. Wenn man das obige Beispiel her nimmt, wird die Variable `temp` erst dann erstellt, wenn ein Wert an die Gruppenadresse gesendet wird. Danach ist die Variable wieder verschwunden und hat somit keinen Wert/Zustand mehr. Das was du benötigst wär ein Debugger und dieser ist nicht so einfach umzusetzen.

Mit Log-Ausgaben à la `println` kommt man schon sehr weit. 

>  Mit Homebridge und Co. ist das ungefähr derselbe Aufwand und willst Du da nicht besser sein? :-)

Naja, die Standardfunktionen von hkknx sind ja ohne Programmieraufwand verfügbar. Für das Erstellen von Automationen wird aber ein gewissen Level an Verständnis benötigt. Ohne dem Verständnis würde man bei grafischen Logik-Editoren auch nicht weit kommen.

- Für das Einsehen von Zuständen der Variablen würde ich `println` empfehlen.
- Wenn es um das Senden von Werten geht, könnte man eventuell beim Busmonitor eine Funktion einbauen, mit der man Werte auf den KNX Bus senden kann. Ähnlich wie bei der ETS.

**Comment by chef76** [comment]
Mit dem Senden von Werten im BusMonitor wäre auf jeden Fall auch schon was erreicht. Wäre schön, wenn Du das einbauen könntest. Danke!
